### DISCUSSION TITLE
Crop to image or crop to an element

### DISCUSSION THREAD

**Discussion by AnReZa** [vvvv/SVG#855] [Q&A]
In Inkscape, you have the option to crop the canvas to either the outer dimensions of your drawing, or to your current selection of elements. Especially the first option would be really handy. Sometimes, my canvas is much bigger than the actual drawing, so an auto-crop function would solve many problems.

Is there already a way to achieve that?

**Comment by AnReZa** [comment]
Any news?

**Comment by H1Gdev** [comment]
@AnReZa 

You can crop with `void Draw(Bitmap bitmap)`.
But it's not an arbitrary place, so it's probably not what you want...(There are also side effects.)

- `Bitmap Draw()`

![_output](https://user-images.githubusercontent.com/11144112/74882162-f2708a80-53b1-11ea-8f8f-869f648501e0.png)

- `void Draw(Bitmap bitmap)` (0, 0, 100, 100)

![output](https://user-images.githubusercontent.com/11144112/74882200-00261000-53b2-11ea-9afa-ec142643df9d.png)

There is a way to crop an arbitrary location by specifying `viewBox`.

- `viewBox` (50, 50, 100, 100)

```cs
            svgDoc.ViewBox = new SvgViewBox()
            {
                MinX = 50,
                MinY = 50,
                Width = 100,
                Height = 100,
            };
            svgDoc.Width = 100;
            svgDoc.Height = 100;
            var image = svgDoc.Draw();
```

![___output](https://user-images.githubusercontent.com/11144112/74882738-f4871900-53b2-11ea-9ab3-a39c082c878e.png)

**Comment by AnReZa** [comment]
@H1Gdev But there is no auto-crop feature or any "crop to specific element" option, right?

**Comment by H1Gdev** [comment]
> But there is no auto-crop feature or any "crop to specific element" option, right?

You may be able to do so(use Graphics, SvgElement, etc.), but we don't provide it as method.

**Comment by AnReZa** [comment]
I see. Thank you so far. The viewbox solution works fine in my case.

**Comment by H1Gdev** [comment]
FYI: Crop using `Graphics`

```cs
using (var image = new Bitmap(100, 100))
{
    using (var g = Graphics.FromImage(image))
    {
        g.TranslateTransform(-50, -50);
        svgDoc.Draw(g);
    }
}
```

This has advantage of not changing the SVG DOM.

**Comment by AnReZa** [comment]
@H1Gdev Thank you very much. In my specific case, I created an image editor using your library to add annonations over photos. The photo is scaled to the user's screen size, but the original aspect ration remains. When rendering the whole thing as JPG, I want to crop the "letterbox" around the background image.

```c#
/// <summary>Internal method to render the SVG as bitmap.</summary>
/// <param name="imgFormat">The rendered image as memory stream.</param>
/// <returns>A bitmap object containing the rendered image.</returns>
private Bitmap RenderAsBitmap(ImageFormat imgFormat = null) {
    if (string.IsNullOrEmpty(Shape))
        throw new InvalidOperationException("This method can only be executed, when the Shape parameter is set.");

    // Replace the links with a Base64 representation
    string svgSrc = Regex.Replace(
        Shape,
        @"(href="")((?>\/QCProcessing)?\/Diagnosis\/[^""]*)("")",
        $"$1data:{MimeMapping.GetMimeMapping(Picture.OriginalFileName)};base64,{Convert.ToBase64String(Picture.Data)}$3"
    );

    // Create SVG object from source code
    SvgDocument svg = SvgDocument.FromSvg<SvgDocument>(svgSrc);

    // Create a Bitmap object from original picture for size comparison
    using Bitmap originalBackgroundImage = GetImageFromByteArray(Picture.Data);

    // Calculate aspect ratios
    float originalAspectRatio = (float)originalBackgroundImage.Width / originalBackgroundImage.Height;
    float currentAspectRatio = svg.Bounds.Width / svg.Bounds.Height;

    // Calculate the actual width of the background image in context to the canvas (due to different aspect ratios of picture and screen)
    float actualBgImgWidth = originalAspectRatio < currentAspectRatio ? svg.Bounds.Width / currentAspectRatio * originalAspectRatio : svg.Bounds.Width;
    float actualBgImgHeight = originalAspectRatio < currentAspectRatio ? svg.Bounds.Height : svg.Bounds.Height / currentAspectRatio * originalAspectRatio;

    // Crop image
    svg.ViewBox = new SvgViewBox(
        (svg.Bounds.Width - actualBgImgWidth) / 2,
        (svg.Bounds.Height - actualBgImgHeight) / 2,
        actualBgImgWidth,
        actualBgImgHeight
    );

    svg.Width = actualBgImgWidth;
    svg.Height = actualBgImgHeight;

    // TODO: Scale image
    //svg.Transforms ??= new Svg.Transforms.SvgTransformCollection();
    //svg.Transforms.Add(new Svg.Transforms.SvgScale(
    //    originalBackgroundImage.Width / actualBgImgWidth,
    //    originalBackgroundImage.Height / actualBgImgHeight
    //));

    // Return a render of the image
    return svg.Draw();
}
```

**Comment by H1Gdev** [comment]
```cs
    svg.Width = actualBgImgWidth;
    svg.Height = actualBgImgHeight;
```
I think you set the width and height to shrink here.

**Comment by AnReZa** [comment]
@H1Gdev So if I set this to a larger number, it would automatically rescale the whole image?

**Comment by H1Gdev** [comment]
- 120x120
```cs
            svgDoc.ViewBox = new SvgViewBox(
                50,
                50,
                100,
                100
            );
            svgDoc.Width = 120;
            svgDoc.Height = 120;
```

![120x120](https://user-images.githubusercontent.com/11144112/75050016-86efff80-550e-11ea-9337-905dda7e1c68.png)

- 80x80
```cs
            svgDoc.ViewBox = new SvgViewBox(
                50,
                50,
                100,
                100
            );
            svgDoc.Width = 80;
            svgDoc.Height = 80;
```
![80x80](https://user-images.githubusercontent.com/11144112/75050058-98390c00-550e-11ea-91b8-1c77bd1f4471.png)
